[Launch Handler article] Update to latest explainer version (#3368)
GoogleChrome/developer.chrome.com: site/en/docs/web-platform/launch-handler/index.md

layout: 'layouts/doc-post.njk'
title: Launch Handler API
description: >
  Launch handler lets you control how your app is launched, for example, whether it uses an existing
  or a new window and whether the chosen window is navigated to the launch URL.
subhead: >
  Control how your app is launched.
date: 2021-12-14
updated: 2022-08-03
authors:
  - thomassteiner
---

The Launch Handler API lets you control how your app is launched, for example, whether it uses an

> existing or a new window and whether the chosen window is navigated to the launch URL. This also
> enqueues a `LaunchParams` object in the launched page's `window.launchQueue`, similar to the File
> Handling API.

Reviewer: ```suggestion
existing or a new window and whether the chosen window is navigated to the launch URL. As with
the File Handing API, this also enqueues a `LaunchParams` object in the launched page's `window.launchQueue` Handling API.
```

    URL.
  - `navigate-existing`: The most recently interacted with browsing context in a web app
    window is navigated to the launch's target URL.
  - `focus-existing`: The most recently interacted with browsing context in a web app window
    is chosen to handle the launch. A new `LaunchParams` object with its `targetURL` set to the
    launch URL will be enqueued in the document's `window.launchQueue`.
  - `auto`: The behavior is up to the user agent to decide what works best for the platform. For
    example, mobile devices only support single clients and would use `existing-client`, while
    desktop devices support multiple windows and would use `navigate-new` to avoid data loss.

The `client_mode` property also accepts a list (array) of values, where the first valid value will be
used. This is to allow new values to be added to the spec without breaking backwards compatibility
with existing implementations.

For example, if the hypothetical value `"matching-url-client"` were added, sites would specify
`"client_mode": ["matching-url-client", "navigate-existing"]` to continue to control the

> behavior of older browsers that did not support `"matching-url-client"`.

Reviewer 2: Tiny nit: I decided to call this `focus-matching-url` in the explainer.

Author: Will change to that then. Thanks!

Author: See https://github.com/GoogleChrome/developer.chrome.com/pull/3378.

</div>

### Try out the Launch Handler API

During the trial phase you can test the API by one of two methods.

#### Local testing

To experiment with the Launch Handler API locally, without an origin trial token, enable the
`#enable-desktop-pwas-launch-handler` flag in `about://flags`.

#### Register for the origin trial

> Starting in Chromium 98, the Launch Handler API will be available as an

Reviewer: ```suggestion
Starting in Chromium 98, the Launch Handler API is available as an
```

[origin trial](/docs/web-platform/origin-trials/) in Chromium. The origin trial is expected to end

> in Chromium 106 (October 26, 2022).

Reviewer: ```suggestion
in Chromium 106 (October 25, 2022).
```

Reviewer: This is the date [on the dashboard](https://developer.chrome.com/origintrials/#/view_trial/2978005253598740481).

[Register here](https://developers.chrome.com/origintrials/#/trials/active).

{% endAside %}

## Interfaces

The Launch Handler API defines two new interfaces.

> `launchParams` : An object containing the targetURL to be handled by the consumer. `launchQueue` :
> Queues launches until they are handled by the specified consumer.

Reviewer: ```suggestion
`LaunchParams` : An object containing the `targetURL` to be handled by the consumer.
`LaunchQueue` : Queues launches until they are handled by the specified consumer.
```

## The `launch_handler` manifest member

To declaratively specify the launch behavior of your app, add the `launch_handler` manifest member
to your manifest. It has one sub-field called `client_mode`. It lets you control whether a new or an

> existing client should be launched and if this client should be navigated. The Web App Manifest
> excerpt below shows a file with exemplary values that would always route all launches to a new

Reviewer: ```suggestion
existing client should be launched and if this client should be navigated. The example
below shows a file with exemplary values that would always route all launches to a new
```

client.

```json
{
  "launch_handler": {
    "client_mode": "navigate-new"
  }
}
```

If unspecified, `launch_handler` defaults to `{"client_mode": "auto"}`. The allowed values for the

> sub-fields are as follows:

Reviewer: ```suggestion
sub-fields are:
```

### Using window.launchQueue

> In the following code, the function `extractSongID()` is used to extract a `songID` from the URL

Reviewer: ```suggestion
In the following code, the function `extractSongID()` extracts a `songID` from the URL
```

> passed on launch. This is then used to play a song in a music player PWA.

Reviewer: ```suggestion
passed on launch. This is used to play a song in a music player PWA.
```
